File: Artificial Intelligence/Lista2/zad1.py
import random
import numpy as np
import math
import time
from collections import deque

ticks = 0
total_time = 0.0

def possible_block_positions(seq_len, block_lens):
    block_num = len(block_lens)
    # queue holdings states. Each state is list of positions of blocks.
    q = deque() 
    pos = 0
    positions = []
    while pos + sum(block_lens) + len(block_lens) - 1 <= seq_len:
        q.append([(pos, block_lens[0])])
        pos += 1
     
    while not (len(q) == 0):
        state = q.pop()
        state_len = len(state)

        if state_len != block_num:
            # new possible position: last block pos + last block len + 1
            pos = state[-1][0] + state[-1][1] + 1
            remaining_pixels = sum(block_lens[state_len:])
            remaining_blocks = len(block_lens[state_len:])
            while pos + remaining_pixels + remaining_blocks - 1 <= seq_len:
                new_state = list(state)
                new_state.append((pos, block_lens[state_len]))
                q.append(new_state)
                pos += 1
        else:
            x = np.zeros(seq_len)
            for s in state:
                x[s[0]:s[0]+s[1]] = 1
            positions.append(x)

    return np.vstack(positions)

def opt_dist(seq, states):
    seq = np.array(seq)
    states = np.copy(states)
    return np.amin(np.abs(states - seq).sum(axis=1))

def random_fill(image):
    for row in image:
        for i in range(len(row)):
            row[i] = random.randint(0,1)
            
def print_to_file(image):
    for row in image:
        signs = []
        for i in row:
            signs.append('#' if i == 1 else '.') 
        f_out.write(''.join(signs) + '\n')

def print_image(image):
    for row in image:
        signs = []
        for i in row:
            signs.append('#' if i == 1 else '.') 
        print(''.join(signs) + '\n')
    print('\n*******\n')

# returns list of rows and columns with invalid blocks
def get_invalid_rows_cols(col_opt_dists, row_opt_dists):
    invalid = []
    for row_num in range(len(row_opt_dists)):
        if row_opt_dists[row_num] > 0:
            invalid.append((row_num, -1))
            
    for col_num in range(len(col_opt_dists)):
        if col_opt_dists[col_num] > 0:
            invalid.append((-1, col_num))    
            
    return invalid

# fixes pixels which are '#' in all possible blocks
def fix_positions(descripts, rows, cols, fix_rows):
    solid = np.ones((rows, cols))
    for idx in range(len(descripts)):
        mat = np.any(descripts[idx] == 0, axis=0)
        if fix_rows:
            solid[idx, mat] = 0
        else:
            solid[mat, idx] = 0
    return solid

def WalkSat(descriptions, draw=False):
    rows, cols = len(descriptions[0]), len(descriptions[1])

    image = np.random.random_integers(0, 1 , size=(rows,cols))

    rows_block_positions = [possible_block_positions(cols, r) 
        for r in descriptions[0]]
    cols_block_positions = [possible_block_positions(rows, c) 
        for c in descriptions[1]]

    fixed_by_row = fix_positions(rows_block_positions, rows, cols, True)
    fixed_by_col = fix_positions(cols_block_positions, rows, cols, False)
    fixed = fixed_by_col + fixed_by_row
    image[fixed > 0] = 1
    
    col_opt_dists = [opt_dist(image[:, i], cols_block_positions[i]) 
        for i in range(cols)]
    row_opt_dists = [opt_dist(image[i, :], rows_block_positions[i]) 
        for i in range(rows)]

    iterations = 0
    while True:
        invalid_rc = get_invalid_rows_cols(col_opt_dists, row_opt_dists)

        if len(invalid_rc) == 0:
            print('found in: ', iterations)
            break

        rnd = random.choice(invalid_rc)
        min_dist, min_dist_pos = 999999, (-1,-1)
        num = 0

        # go through all pixels in row/column
        while num < (rows if rnd[0] == -1 else cols): 
            row_num = rnd[0] if rnd[1] == -1 else num
            col_num = rnd[1] if rnd[0] == -1 else num

            if fixed[row_num, col_num] > 0: 
                num += 1
                continue

            row = image[row_num,:].tolist()
            col = image[:, col_num].tolist()
            old_row_dist = row_opt_dists[row_num]
            old_col_dist = col_opt_dists[col_num]
            old_dist = old_row_dist + old_col_dist

            col[row_num] = 1 - col[row_num] 
            row[col_num] = 1 - row[col_num]

            new_row_dist = opt_dist(row, rows_block_positions[row_num])
            new_col_dist = opt_dist(col, cols_block_positions[col_num])
            dist = new_row_dist + new_col_dist

            if dist - old_dist == min_dist and old_row_dist > new_row_dist \
                and old_col_dist > new_col_dist and random.random() > 0.5:
                min_dist_pos = (row_num, col_num)
            if dist - old_dist < min_dist and old_row_dist >= new_row_dist \
                and old_col_dist >= new_col_dist:
                min_dist_pos = (row_num, col_num)
                min_dist = dist - old_dist
            elif dist - old_dist < min_dist and np.random.rand() < 0.20: # 0.09
                min_dist_pos = (row_num, col_num)
                min_dist = dist - old_dist
            
            num += 1

        if min_dist_pos == (-1,-1): continue

        # negate pixel
        image[min_dist_pos] = 1 - image[min_dist_pos]

        # recompute opt dists
        idx = min_dist_pos[0]
        row_opt_dists[idx] = opt_dist(image[idx, :], rows_block_positions[idx])
        idx = min_dist_pos[1]
        col_opt_dists[idx] = opt_dist(image[:, idx], cols_block_positions[idx])
        
        iterations += 1
        if iterations > rows*cols*200:
            print_image(image)
            image = np.random.random_integers(0, 1 , size=(rows,cols))
            image[fixed > 0] = 1
            iterations = 0
            
    print_to_file(image)


f = open("zad_input.txt", 'r')
f_out = open("zad_output.txt", 'w')

row = []
col = []
for i,l in enumerate(f):
    l = l.split()

    if i == 0:
        row_num = int(l[0])
        col_num = int(l[1])
    elif len(row) < row_num:
        row.append([int(x) for x in l])
    else:
        col.append([int(x) for x in l])


WalkSat((row,col))

File: Artificial Intelligence/Lista2/zad1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;

namespace AILista2
{
	class MainClass
	{
		List<List<Tuple<int, int>>> possible_block_positions(int seq_len, List<int> block_lens)
		{
			int block_num = block_lens.Count;
			Queue<List<Tuple<int, int>>> q = new Queue<List<Tuple<int, int>>>();
			int pos = 0;
			var positions = new List<List<Tuple<int, int>>> ();

			while(pos + block_lens.Sum () +  block_lens.Count - 1 <= seq_len )
			{
				q.Enqueue (new List<Tuple<int, int>>() { new Tuple<int,int> (pos, block_lens[0]) });
				pos += 1;
			}

			while (q.Count != 0)
			{
				var state = q.Dequeue ();
				int state_len = state.Count;

				if (state_len != block_num) {
					pos = state.Last ().Item1 + state.Last ().Item2 + 1;
					while (pos + block_lens.Skip (state_len).Sum () + block_lens.Skip (state_len).Count () - 1 <= seq_len) {
						var new_state = state.ToList ();
						new_state.Add (new Tuple<int, int>(pos, block_lens[state_len]));
						q.Enqueue (new_state);
						pos += 1;
					}
				}
				else
				{
					positions.Add (state);
				}
			}

			return positions;

		}

		int opt_dist(int[] seq, Tuple<int, int>[][] states)
		{
			int min_swaps = 9999;
			int seq_sum = seq.Sum ();
			int block_sum = 0;
			for (int i = 0; i < states.ElementAt (0).Count (); i++)
				block_sum += states.ElementAt (0).ElementAt (i).Item2;

			foreach (var state in states) {
				int swaps = 0;
				foreach (var s in state) {
					int sum = 0;
					for (int i = s.Item1; i < s.Item1 + s.Item2; i++)
						sum += seq [i];
					swaps += s.Item2 - sum;
				}

				swaps += seq_sum + swaps - block_sum;
				min_swaps = swaps < min_swaps ? swaps : min_swaps;
				if (min_swaps == 0)
					return 0;
			}
			return min_swaps;
		}

		void random_fill(int[,] image)
		{
			Random rnd = new Random();
			for(int x=0; x<image.GetLength (0); x++)
				for(int y=0; y<image.GetLength (1); y++)
					image[x,y] = rnd.Next (0,2);
		}

		void draw_image(int[,] image)
		{
			StringBuilder builder = new StringBuilder();
			for (int x = 0; x < image.GetLength (0); x++) {
				for (int y = 0; y < image.GetLength (1); y++) {
					builder.Append (image[x,y] == 1? '#' : '.');
				}
				builder.AppendLine ();
			}
			System.IO.StreamWriter file = new System.IO.StreamWriter(@"zad_output.txt");
			file.Write(builder.ToString());
			file.Close ();
		}
			
		void print_image(int[,] image)
		{
			StringBuilder builder = new StringBuilder();
			for (int x = 0; x < image.GetLength (0); x++) {
				for (int y = 0; y < image.GetLength (1); y++) {
					builder.Append (image[x,y] == 1? '#' : '.');
				}
				builder.AppendLine ();
			}
			//Console.Write(builder.ToString()); 
		}

		List<Tuple<int,int>> get_invalid_rows_cols(int[] col_opt_dists, int[] row_opt_dists)
		{
			var invalid = new List<Tuple<int,int>> ();

			for (int col_num = 0; col_num < col_opt_dists.Length; col_num++) {
				if (col_opt_dists [col_num] > 0) {
					invalid.Add (new Tuple<int, int> (-1, col_num));
				}
			}

			for (int row_num = 0; row_num < row_opt_dists.Length; row_num++) {
				if (row_opt_dists [row_num] > 0) {
					invalid.Add (new Tuple<int, int> (row_num, -1));
				}
			}

			return invalid;
		}

		int[] getColumn(int[,] image, int col_num)
		{
			int[] column = new int[image.GetLength (0)];
			for (int row = 0; row < image.GetLength (0); row++)
				column [row] = image [row, col_num];

			return column;
		}

		int[] getRow(int[,] image, int row_num)
		{
			int[] row = new int[image.GetLength (1)];
			for (int col = 0; col < image.GetLength (1); col++)
				row [col] = image [row_num, col];

			return row;
		}
			
		void WalkSat(List<List<int>> row_descriptions, List<List<int>> col_descriptions)
		{
			Random rnd = new Random ();
			int rows = row_descriptions.Count;
			int cols = col_descriptions.Count;

			int[,] image = new int[rows, cols];
			random_fill (image);

			var rows_block_positions = new List<Tuple<int, int>[][]> ();
			var cols_block_positions = new List<Tuple<int, int>[][]> ();
			foreach(List<int> row_desc in row_descriptions)
			{
				var b = possible_block_positions (cols, row_desc);
				Tuple<int, int>[][] r = b.Select (s => s.ToArray ()).ToArray ();//new Tuple<int,int>[b.Count, b[0].Count];
				rows_block_positions.Add (r);
			}

			foreach(List<int> col_desc in col_descriptions)
			{
				var b = possible_block_positions (rows, col_desc);
				Tuple<int, int>[][] c = b.Select (s => s.ToArray ()).ToArray ();
				cols_block_positions.Add (c);
			}

			int[] col_opt_dists = new int[cols];
			for (int i = 0; i < cols; i++)
				col_opt_dists[i] = opt_dist (getColumn(image, i), cols_block_positions[i]);
			int[] row_opt_dists = new int[rows];
			for (int i = 0; i < rows; i++)
				row_opt_dists[i] = opt_dist (getRow (image, i), rows_block_positions[i]);
			
			int iterations = 0;

			while (true) {
				//print_image (image);
				List<Tuple<int,int>> invalid_rc = get_invalid_rows_cols (col_opt_dists, row_opt_dists);
				if (invalid_rc.Count == 0) {
					Console.WriteLine ("found in:" + iterations.ToString ());
					print_image (image);
					break;
				}
				Tuple<int, int> selected = invalid_rc[rnd.Next (0, invalid_rc.Count)];

				int min_dist = 99999;
				Tuple<int, int> min_dist_pos = new Tuple<int, int>(-1, -1);

				int num = 0;
				while (num < (selected.Item1 == -1 ? rows : cols)) {
					int row_num = selected.Item2== -1 ? selected.Item1 : num;
					int col_num = selected.Item1 == -1 ? selected.Item2 : num;

					int[] row = getRow (image, row_num);
					int[] col = getColumn (image, col_num);
					int old_row_dist = row_opt_dists [row_num];
					int old_col_dist = col_opt_dists [col_num];
					int orig_dist = old_col_dist + old_row_dist;

					col [row_num] = (col [row_num] + 1) % 2;
					row [col_num] = (row [col_num] + 1) % 2;
					int new_row_dist = opt_dist (row, rows_block_positions [row_num]);
					int new_col_dist = opt_dist (col, cols_block_positions [col_num]);
					int dist = new_row_dist + new_col_dist;

					if (dist - orig_dist == min_dist && old_row_dist > new_row_dist && old_col_dist > new_col_dist && rnd.NextDouble () < 0.5f) {
						min_dist_pos = new Tuple<int,int> (row_num, col_num);
					} else if (dist - orig_dist < min_dist && old_row_dist >= new_row_dist && old_col_dist >= new_col_dist) {
						min_dist_pos = new Tuple<int,int> (row_num, col_num);
						min_dist = dist - orig_dist;
					} else if (dist - orig_dist < min_dist && rnd.NextDouble () < 0.15f) {
						min_dist_pos = new Tuple<int,int> (row_num, col_num);
						min_dist = dist - orig_dist;
					
					}
					num += 1;
				}
				int x = min_dist_pos.Item1; int y = min_dist_pos.Item2;
				if (x == -1 && y == -1) {
					continue;
				}
				image [x, y] = (image [x, y] + 1) % 2;
				row_opt_dists [x] = opt_dist (getRow (image, x), rows_block_positions [x]);
				col_opt_dists [y] = opt_dist (getColumn (image, y), cols_block_positions [y]);

				iterations += 1;
				if (iterations % 1000 == 0)
					Console.WriteLine (iterations);
				if (iterations > rows * cols * 200) {
					print_image (image);
					random_fill (image);
					iterations = 0;
				}
			}
			draw_image (image);


		}
		public static void Main (string[] args)
		{
			System.IO.StreamReader file = new System.IO.StreamReader(@"zad_input.txt");

			List<List<int>> row = new List<List<int>> ();
			List<List<int>> col = new List<List<int>> ();
			int row_num = 0;
			string line = "";
			int idx = 0;
			while (!file.EndOfStream) {
				line = file.ReadLine ();
				string[] nums = line.Split (new Char[] { ' ' });

				if (idx == 0) {
					row_num = int.Parse (nums [0]);
				} else if (row.Count < row_num) {
					row.Add (nums.Select (s => int.Parse (s)).ToList ());
				} else {
					col.Add (nums.Select(s=> int.Parse (s)).ToList ());
				}

				idx += 1;
			}
				
			new MainClass().WalkSat (row, col);
		}
	}
}
